src/dune_rules/dep_conf_eval.ml

```diff
@@ -126,28 +126,48 @@ let dep expander = function
       (let+ () = Action_builder.dep (Dep.sandbox_config sandbox_config) in
        [])
 
-let make_interpreter ~f ~expander l =
-  Expander.expand_deps_like_field expander ~dep_kind:Optional
-    ~f:(fun expander ->
-      match Result.List.map l ~f:(f expander) with
-      | Ok deps ->
-        let+ l = Action_builder.all deps in
-        List.concat l
-      | Error exn -> Action_builder.fail { fail = (fun () -> reraise exn) })
+let dep expander x = Action_builder.of_result (dep expander x)
+
+let prepare_expander expander =
+  let expander = Expander.set_dep_kind expander Optional in
+  Expander.set_expanding_what expander Deps_like_field
 
 let unnamed ~expander l =
-  let+ _paths = make_interpreter ~f:dep ~expander l in
-  ()
+  let expander = prepare_expander expander in
+  List.fold_left l ~init:(Action_builder.return ()) ~f:(fun acc x ->
+      let+ () = acc
+      and+ _x = dep expander x in
+      ())
 
-let named =
-  make_interpreter ~f:(fun expander -> function
-    | Bindings.Unnamed p ->
-      dep expander p
-      |> Result.map ~f:(fun l ->
-             let+ l = l in
-             List.map l ~f:(fun x -> Bindings.Unnamed x))
-    | Named (s, ps) ->
-      Result.List.map ps ~f:(dep expander)
-      |> Result.map ~f:(fun xs ->
-             let+ l = Action_builder.all xs in
-             [ Bindings.Named (s, List.concat l) ]))
+let named ~expander l =
+  let builders, bindings =
+    let expander = prepare_expander expander in
+    List.fold_left l ~init:([], Pform.Map.empty)
+      ~f:(fun (builders, bindings) x ->
+        match x with
+        | Bindings.Unnamed x -> (dep expander x :: builders, bindings)
+        | Named (name, x) ->
+          let x =
+            Action_builder.memoize ("dep " ^ name)
+              (let+ l = Action_builder.all (List.map x ~f:(dep expander)) in
+               List.concat l)
+          in
+          let bindings =
+            Pform.Map.set bindings (Var (User_var name))
+              (let+ paths = x in
+               Dune_util.Value.L.paths paths)
+          in
+          (x :: builders, bindings))
+  in
+  let builder =
+    let+ l = Action_builder.all (List.rev builders) in
+    Dune_util.Value.L.paths (List.concat l)
+  in
+  let builder = Action_builder.memoize "deps" builder in
+  let bindings = Pform.Map.set bindings (Var Deps) builder in
+  let expander = Expander.add_bindings_full expander ~bindings in
+  let builder =
+    let+ _l = builder in
```

Comment:
Perhaps `Build.ignore` can be used here.

Reply:
Indeed, good catch